### PR TITLE
feature/design of work request actions

### PR DESCRIPTION
**Pull request by szirbucz**

### CODE REVIEW

**Comment by magwas** [comment]
@magwas

File: tidyup.zenta [hunk 1]
```diff
@@ -2565,58 +2837,159 @@ They are brownish in color, to make it </documentation>
...
+          <element xsi:type="zenta:BasicRelationship" id="6ae58303-82eb-438a-9705-56c84f7e63a6" ancestor="36b8d926-8f1a-4aba-805f-4b8ab5696360" source="0a3346bc-64e7-46e7-85ae-e621c92dd46f" target="0fe9f41f-fe11-4c5f-a787-6bd58e29cd4b"/>
+          <element xsi:type="zenta:BasicRelationship" id="1d306fa7-81a9-475d-8063-fe6c5b44a525" ancestor="c2720c1a-1642-416d-802b-07fe168edfb6" source="0fe9f41f-fe11-4c5f-a787-6bd58e29cd4b" target="fa075959-26f0-4dbf-b7ad-2e755911964a"/>
+          <element xsi:type="zenta:BasicRelationship" id="dff6f982-e9fb-4c30-be3f-18aa43994315" ancestor="c2720c1a-1642-416d-802b-07fe168edfb6" source="a70d1022-cca6-47fa-bc98-42791f18f35b" target="40c4dd90-cc92-4184-9e67-a548b63352ee"/>
+          <element xsi:type="zenta:BasicRelationship" id="61cbeddf-512f-49ad-af85-b9293c4f3e06" ancestor="c2720c1a-1642-416d-802b-07fe168edfb6" source="354a621a-94c4-4184-8401-efffc6797178" target="40c4dd90-cc92-4184-9e67-a548b63352ee"/>
+        </folder>
+      </folder>
       <element xsi:type="zenta:ZentaDiagramModel" name="Get work requests" id="7d85e937-a3af-4621-a9a8-f8ce5cf35b46" connectionRouterType="1">
```
Comment (magwas) [inline tidyup.zenta]: We should totally discuss this

File: tidyup.zenta [hunk 2]
```diff
@@ -2706,18 +3078,21 @@ They are brownish in color, to make it </documentation>
...
+      <element xsi:type="zenta:BasicRelationship" id="d92300f4-cdef-4260-9dac-49f7a1de4f76" ancestor="be55bb68-44ee-4506-bcb3-ac7b86761efa" source="100db000-a399-47bc-9374-31b0c690243a" target="40c4dd90-cc92-4184-9e67-a548b63352ee"/>
+      <element xsi:type="zenta:BasicRelationship" id="61724710-dd86-4eec-93cc-bdf31951e5f3" ancestor="8be1389f-6b09-4a24-a853-c71a09f81a5e" source="282a7d3f-2eca-4fe0-ad1b-527d2e918c0c" target="40c4dd90-cc92-4184-9e67-a548b63352ee"/>
     </folder>
-    <folder name="4.3 Open own work request" id="54cf6746-278f-488a-a66d-d510b257fe01">
-      <element xsi:type="zenta:ZentaDiagramModel" name="Customeropen work request" id="e7f94aaf-4ae2-42b5-b90a-dfdcddda06c4" connectionRouterType="1">
+    <folder name="4.3 Open work request" id="54cf6746-278f-488a-a66d-d510b257fe01">
+      <element xsi:type="zenta:ZentaDiagramModel" name="Open work request" id="e7f94aaf-4ae2-42b5-b90a-dfdcddda06c4" connectionRouterType="1">
```
Comment (magwas) [inline tidyup.zenta]: I have to understand the parameter-type-pathParameter pattern. I feel some redundancy here.

File: tidyup.zenta [hunk 3]
```diff
@@ -2764,29 +3144,309 @@ They are brownish in color, to make it </documentation>
...
+      <element xsi:type="zenta:BasicRelationship" id="984ea987-9fba-4527-8da7-a4abfd327e64" ancestor="a99c1f63-c4f6-4d0a-b9e8-5bf3aa5284a2" source="8f4c4d96-92aa-4afb-b7d0-cf69f4c97278" target="172104e1-082f-40f6-be8a-0c1ea1c633e4"/>
+      <element xsi:type="zenta:BasicRelationship" id="1838b5a3-ae9f-4cc1-b38a-7d021e7273f5" ancestor="a99c1f63-c4f6-4d0a-b9e8-5bf3aa5284a2" source="8f4c4d96-92aa-4afb-b7d0-cf69f4c97278" target="37405dc7-3a2d-4f80-818f-3141a36e5473"/>
+      <element xsi:type="zenta:BasicRelationship" id="fdeac8a0-76b3-40c3-8e4a-881a55efa0ae" ancestor="a99c1f63-c4f6-4d0a-b9e8-5bf3aa5284a2" source="8f4c4d96-92aa-4afb-b7d0-cf69f4c97278" target="018d7b46-7a73-480d-a169-946cae4bfd70"/>
+      <element xsi:type="zenta:BasicRelationship" id="bdbcbd11-dbcc-4f52-856d-6279c9ed5d4e" ancestor="3dd7b0b1-bd9f-4a13-b5d9-9db0200af5b8" source="dbc84133-b122-4787-8d38-6d0328c4ba53" target="1f8ba2a4-45ab-44d5-b05f-ec8c513e8506"/>
+    </folder>
+    <folder name="4.4 Revoke work request" id="a1d14a5b-b1ec-431d-99b2-2172a988e90b">
+      <element xsi:type="zenta:ZentaDiagramModel" name="Revoke work request" id="8404e64e-f8f8-4777-9fec-34cc413cf2db" connectionRouterType="1">
```
Comment (magwas) [inline tidyup.zenta]: Unnecessary and not greyed out elements

File: tidyup.zenta [hunk 4]
```diff
@@ -2764,29 +3144,309 @@ They are brownish in color, to make it </documentation>
...
+      <element xsi:type="zenta:BasicRelationship" id="cc193037-d5ae-4859-9e87-aa3c63b467a0" ancestor="a0ec85e2-1b1a-47a4-8c98-9f413bb7b995" source="01de6a91-1b89-40a1-9022-df6f6ee69cbe" target="50a2a981-62a5-4ef2-bccf-0da570139558"/>
+      <element xsi:type="zenta:BasicRelationship" id="c502a509-55cb-49d7-a6e1-48aae2f38d9e" ancestor="37300370-a7f6-47a8-9cc8-e718c48b8d3d" source="61d29545-19bb-4b2f-9bf0-b6dbdf9b2b1b" target="d1d4d170-1546-402b-a51f-aa05315cf646"/>
+      <element xsi:type="zenta:BasicRelationship" id="3dfbeb30-cf8e-470f-88da-64122a58dc58" ancestor="e3ad9158-0def-47d6-a531-196ad3733004" source="b8b44878-f2fa-4fa3-bd01-a6e51538ad96" target="562506d7-6287-43dd-a9ab-daeca22851d5"/>
+      <element xsi:type="zenta:BasicRelationship" id="b9264289-3aa6-4ff3-aa97-a665f20bd237" ancestor="e3ad9158-0def-47d6-a531-196ad3733004" source="d967d3e1-f623-4e10-a784-4160fc470d98" target="562506d7-6287-43dd-a9ab-daeca22851d5"/>
+    </folder>
+    <folder name="4.5 Find work request" id="f6b4c552-4b92-4707-afe7-df89b96dc531">
+      <element xsi:type="zenta:ZentaDiagramModel" name="Find Work Request" id="0488554b-1479-44b5-92a4-b62af134dc88" connectionRouterType="1">
```
Comment (magwas) [inline tidyup.zenta]: unnecessary and not greyed out elements

**Comment by magwas** [comment]
there are also model errors.

File: tidyup.zenta [hunk 5]
```diff
@@ -2499,294 +2568,1122 @@ They are brownish in color, to make it </documentation>
   </folder>
   <folder name="4. Software Design" id="620f8ada-d4be-4595-b6a8-f1c480321d3e">
     <folder name="4.1 Create Work Request" id="3f2e8196-9760-46ee-8f26-5f277758cea9">
-      <element xsi:type="zenta:ZentaDiagramModel" name="Create work request" id="48badaea-2f4e-4cac-b1a1-0fdb37978642" connectionRouterType="1">
```
Comment (magwas) [inline tidyup.zenta]: It is good practice to keep the direction of a tree, and it be either top-down or left-right. You started top-down and continued left-right.

File: tidyup.zenta [hunk 6]
```diff
@@ -2499,294 +2568,1122 @@ They are brownish in color, to make it </documentation>
...
-        <child xsi:type="zenta:DiagramObject" id="14beb458-cc25-4b41-b49a-d9747d0ad0f6" targetConnections="584e5f8f-770d-44e7-8d32-4ceece60e6ae 416d1f24-9f41-4db5-a0a7-d526df9fba2f" zentaElement="247aa06e-5547-4eca-8078-9e6c21303bab">
```
Comment (magwas) [inline tidyup.zenta]: If the order is important we are talking about Process Steps, if not, then those are Behaviours.

**Comment by magwas** [comment]
No behaviours on the CreateWorkRequestController view. The dependent controllers are shown, but the process steps belonging to them are not. How the coder should figure out what the order of steps should be?

**Comment by magwas** [comment]
The WorkRequestListDTO view does not show WorkRequestListDTO

**Comment by magwas** [comment]
The parameter layout in FindWorkRequestService is interesting. We should discuss whether we mandate one or allow both.

**Comment by magwas** [comment]
How we know which work request we are talking about in WorkRequestRepository.setInactive ?

**Comment by magwas** [comment]
It is also important which folder we put which elements in. As a rule of thumb, the elements should appear near the view first mentioning them. On the long term we should have a standard for each type of views, which tells what element should and can be on them, what is the overall layout, which elements should be greyed out, and which elements should be in the same directory.

File: tidyup.zenta [hunk 7]
```diff
@@ -2489,304 +2543,1211 @@ They are brownish in color, to make it </documentation>
...
+          </element>
+          <element xsi:type="zenta:BasicObject" name="Address" id="2484c0a7-c78c-46bf-9b8f-0811054e3e9f" ancestor="41b4d079-dbd5-4df2-91db-6a596661ab0d"/>
+          <element xsi:type="zenta:BasicObject" name="id" id="046ab36a-3238-45a1-8762-0899f280dc51" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="country" id="547e922b-64fc-4526-b437-2e8e0ed91097" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="city" id="f3cee831-6fe6-4f97-89f3-e02ec798ac78" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="address" id="b3ae2ead-2da8-4165-a951-c6cb3eccbaa5" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+        </folder>
+      </folder>
+      <folder name="1. Presentation Layer" id="82bccda3-c8b2-4f0e-9a06-dee605ee9371">
+        <element xsi:type="zenta:ZentaDiagramModel" name="CreateWorkRequestController" id="3c59b07f-3960-44c1-8815-9fd3684b1ba4" connectionRouterType="1">
```
Comment (magwas) [inline tidyup.zenta]: I see the dependent controllers greyed out, but cannot see the contained process steps according to them.
The order of the steps are not apparent from the view this way.
(Yes, it is trivial, but if you put up something for guessing, some will guess wrong invariably).

Comment (magwas) [inline tidyup.zenta]: The '/' in relation names is to tell how their should be spelled out in one or other direction.
"implements is implemented by" clearly needs such thing, and it seems that there is no constraint related to that, which is strange.

File: tidyup.zenta [hunk 8]
```diff
@@ -2489,304 +2543,1211 @@ They are brownish in color, to make it </documentation>
...
+            <bounds x="264" y="36" width="144" height="73"/>
+            <sourceConnection xsi:type="zenta:Connection" id="fec5997b-bad9-4084-8a3b-33d80e3c4489" source="aee055bb-cd3e-41f9-8db8-b12b09f71537" target="2469106d-d41b-47a1-beff-7c6a974437b3" relationship="63adbb25-ebcd-47d3-9ebf-02da3e49a920"/>
+          </child>
+        </element>
+        <element xsi:type="zenta:BasicObject" name="/work-request" id="98d26fdd-61da-4acc-a00c-4e9b056c99be" ancestor="2cb904ef-0cda-422b-82f6-0509c595b8fd"/>
+        <element xsi:type="zenta:BasicObject" name="CreateWorkRequestControler" id="c344d36f-3f83-477e-8921-f9242185d883" ancestor="5491f72b-93ec-4ad5-b6a9-2653fe808a03"/>
+        <element xsi:type="zenta:BasicRelationship" id="63adbb25-ebcd-47d3-9ebf-02da3e49a920" ancestor="3dd7b0b1-bd9f-4a13-b5d9-9db0200af5b8" source="e5a09833-8e15-4ad8-8e84-7fae3eff5e05" target="c344d36f-3f83-477e-8921-f9242185d883"/>
+      </folder>
+      <folder name="2. Business Layer" id="db99bafc-1e9f-4bde-b8aa-3f11d7af54ba">
+        <element xsi:type="zenta:ZentaDiagramModel" name="CreateWorkRequestService" id="061b73b0-3f7d-4656-8f8f-e64d806582c9">
```
Comment (magwas) [inline tidyup.zenta]: It creates a Work request entity. But what if the address does not yet exist? Error is signaled, or the address is stored?

File: tidyup.zenta [hunk 9]
```diff
@@ -2489,304 +2543,1211 @@ They are brownish in color, to make it </documentation>
...
+            <bounds x="507" y="120" width="-1" height="-1"/>
+            <sourceConnection xsi:type="zenta:Connection" id="f89a06af-3769-4810-8950-069a9978bdcf" source="8ad23b96-ea0f-488e-8bc3-6d10454b933a" target="9b3ba021-6318-48fb-ae61-46e16a3c4118" relationship="15bf6085-2890-4c80-b8cd-ad530ff8cb55"/>
+            <sourceConnection xsi:type="zenta:Connection" id="d8e724f0-9dc3-43a2-b1bb-200fde5b8e68" source="8ad23b96-ea0f-488e-8bc3-6d10454b933a" target="33b8c7a0-f161-481c-9a9b-b7933f9f1690" relationship="ea801304-9ea9-4595-bddb-095804e7f799"/>
+          </child>
+          <child xsi:type="zenta:DiagramObject" id="2c009986-fee7-45a8-a353-1793f51c7eab" zentaElement="440f2e73-5c11-4b5f-b81f-bb8bcf092277">
+            <bounds x="507" y="24" width="-1" height="-1"/>
+            <sourceConnection xsi:type="zenta:Connection" id="584ba308-681a-4355-a570-9b50b8636559" source="2c009986-fee7-45a8-a353-1793f51c7eab" target="8ad23b96-ea0f-488e-8bc3-6d10454b933a" relationship="95867fb8-31d5-4f08-8c4b-044bc22a3696"/>
+          </child>
+        </element>
+        <element xsi:type="zenta:ZentaDiagramModel" name="CreateWorkRequestValidator" id="52b1d8d9-ff84-4b77-b882-1919b2426747">
```
Comment (magwas) [inline tidyup.zenta]: The substeps are more look like behaviour than process steps, as there is no order between them.
There is a depended service shown without the corresponding depended process step.
Also, I would put the depended process step below the depending one.

File: tidyup.zenta [hunk 10]
```diff
@@ -2489,304 +2543,1211 @@ They are brownish in color, to make it </documentation>
...
+            <sourceConnection xsi:type="zenta:Connection" id="f373dbb2-cee4-418a-82a0-fc8f09b3f77a" source="b2c6e05c-a219-468a-897c-cbd751755c44" target="25c15698-e2d8-43c6-8a85-05a1fe2c53a2" relationship="2633081c-6bdf-43e4-9f3b-d8b8011f3130"/>
+            <sourceConnection xsi:type="zenta:Connection" id="6395816a-e3b5-4317-8b68-6f2c81852d46" source="b2c6e05c-a219-468a-897c-cbd751755c44" target="f3c2ae00-d390-4831-b1a6-a8056b8b2741" relationship="ff746ab2-6743-4ff8-b864-e8cee2f762b6"/>
+            <sourceConnection xsi:type="zenta:Connection" id="f5f45ded-e04b-4276-99ca-40fabad4b424" source="b2c6e05c-a219-468a-897c-cbd751755c44" target="59c00552-c54a-4165-af67-bf79b16194ea" relationship="0cc4fc8a-24b2-4c90-91fb-62a6a6abaa7c"/>
+          </child>
+          <child xsi:type="zenta:DiagramObject" id="f3c2ae00-d390-4831-b1a6-a8056b8b2741" targetConnections="6395816a-e3b5-4317-8b68-6f2c81852d46" zentaElement="713b7ae4-2a40-4660-8178-856e20ba60c8">
+            <bounds x="120" y="148" width="-1" height="-1"/>
+            <sourceConnection xsi:type="zenta:Connection"
```
Comment (magwas) [inline tidyup.zenta]: There is no WorkRequestListDTO shown in the view.

Comment (magwas) [inline tidyup.zenta]: If we would have a WorkrequestListDTO, how would it be different from WorkRequestListEntity?

Comment (magwas) [inline tidyup.zenta]: ownerId is actually a parameter for the service. Maybe there should be some derived relations stating that facts?
The problem is that there is no parameter entity existing for that. (which can be remediated by having derived entities as well)
depending service but no depending process step shown

Comment (magwas) [inline tidyup.zenta]: "CustomerGetWorkRequestsService implements is implemented by Get work requests by customerID" is here twice.
Should have triggered a model error.

Comment (magwas) [inline tidyup.zenta]: The arrows between CustomerGetWorkRequestsService and WorkRequestListEntity are hard to understand. Maybe drawing them apart from each other would help.

Comment (magwas) [inline tidyup.zenta]: depended on service shown but depended on process step not.

Comment (magwas) [inline tidyup.zenta]: Should we have a dedicated relation for "returning a collection of entities served by the repository"?

File: tidyup.zenta [hunk 11]
```diff
@@ -2489,304 +2543,1211 @@ They are brownish in color, to make it </documentation>
...
+          <element xsi:type="zenta:BasicObject" name="workType" id="416fd5e4-674a-4c24-a6b1-b96cd3b72db9" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="country" id="678f6612-31e9-4bd3-bb73-77e1fe0982ab" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="city" id="ceea2751-38e4-4e6a-844e-409499ab5e83" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="address" id="680ce87d-47bf-45c7-b739-7b04dbc28d66" ancestor="b7ee463c-0e72-462b-987d-a7583cef92f4"/>
+          <element xsi:type="zenta:BasicObject" name="WorkType" id="8361ec48-3321-4ebf-a8b1-d78a5642d6c2" ancestor="4567a58e-cb44-4fe4-a082-25460d8b2622"/>
+          <element xsi:type="zenta:BasicObject" name="CreateWorkRequestDTO" id="6b7fdebd-1943-40d4-b5d8-cc36547970bd" ancestor="3e2d26e1-ce3a-4bcb-af15-c159dce65612"/>
+          <element xsi:type="zenta:BasicObject" name="AddressDTO" id="fa075959-26f0-4dbf-b7ad-2e755911964a" ancestor="3e2d26e1-ce3a-4bcb-af15-c159dce65612"/>
```
[the remark this answers is not in the document]
Comment (szirbucz) [reply tidyup.zenta]: Discussed on Skype. We need DTO for separating concerns and hide the backend implementation from the frontend.

File: tidyup.zenta (hunk 7 above)
Comment (szirbucz) [reply tidyup.zenta]: OK

Comment (szirbucz) [reply tidyup.zenta]: I can't see which relation is the target of your concern here.

File: tidyup.zenta (hunk 8 above)
Comment (szirbucz) [reply tidyup.zenta]: Address is stored. I will model it more detailed.

**Pull request closed without merging**